src/iir-susceptibility.cpp

```diff
@@ -0,0 +1,417 @@
+    if (d->P[c][cmp]) {
+      dnew->P[c][cmp] = P;
+      P += ntot;
+      dnew->storage[c][cmp] = P;
+      P += ntot * (D_z-1);
+    }
+  }
+  return (void *)dnew;
+}
+
+#define SWAP(t, a, b)                                                                              \
+  {                                                                                                \
+    t SWAP_temp = a;                                                                               \
+    a = b;                                                                                         \
+    b = SWAP_temp;                                                                                 \
+  }
+
+// stable averaging of offdiagonal components
+#define OFFDIAG(u, g, sx, s)                                                                       \
+  (0.25 * ((g[i] + g[i - sx]) * u[i] + (g[i + s] + g[(i + s) - sx]) * u[i + s]))
+
+void iir_susceptibility::update_P(realnum *W[NUM_FIELD_COMPONENTS][2],
+                                         realnum *W_prev[NUM_FIELD_COMPONENTS][2], double dt,
+                                         const grid_volume &gv, void *P_internal_data) const {
+  iir_data *d = (iir_data *)P_internal_data;
+  size_t ntot = d->ntot;
+  (void)dt;
+  (void)W_prev; // unused;
+
+  FOR_COMPONENTS(c) DOCMP2 {
+    if (d->P[c][cmp]) {
+      const realnum *w = W[c][cmp], *s = sigma[c][component_direction(c)];
+      if (w && s) {
+        realnum *p = d->P[c][cmp], *storage = d->storage[c][cmp];
+
+        // directions/strides for offdiagonal terms, similar to update_eh
+        const direction d = component_direction(c);
+        const ptrdiff_t is = gv.stride(d) * (is_magnetic(c) ? -1 : +1);
+        direction d1 = cycle_direction(gv.dim, d, 1);
+        component c1 = direction_component(c, d1);
+        ptrdiff_t is1 = gv.stride(d1) * (is_magnetic(c) ? -1 : +1);
+        const realnum *w1 = W[c1][cmp];
+        const realnum *s1 = w1 ? sigma[c][d1] : NULL;
+        direction d2 = cycle_direction(gv.dim, d, 2);
+        component c2 = direction_component(c, d2);
+        ptrdiff_t is2 = gv.stride(d2) * (is_magnetic(c) ? -1 : +1);
+        const realnum *w2 = W[c2][cmp];
+        const realnum *s2 = w2 ? sigma[c][d2] : NULL;
+
+        LOOP_OVER_VOL_OWNED(gv, c, i) {
+          if (s[i] != 0) {
+            realnum w_cur;
+            if (s2 && !s1) { // make s1 the non-NULL one if possible
+              SWAP(direction, d1, d2);
+              SWAP(component, c1, c2);
+              SWAP(ptrdiff_t, is1, is2);
+              SWAP(const realnum *, w1, w2);
+              SWAP(const realnum *, s1, s2);
+            }
+            if (s1 && s2) { // 3x3 anisotropic
+              w_cur = s[i] * w[i] + OFFDIAG(s1, w1, is1, is) + OFFDIAG(s2, w2, is2, is);
+            }else if (s1) { // 2x2
+              w_cur = s[i] * w[i] + OFFDIAG(s1, w1, is1, is);
+            }else{ // isotropic
+              w_cur = s[i] * w[i];
+            }
```

Comment:
try replacing `w_cur` with white noise to see if this filter implementation still blows up.